### PR TITLE
test.py: limit cluster pool size to cpu cores available

### PR DESCRIPTION
**Pull request by alecco**
Use at most one core per Scylla process for `PythonTestSuite` and `TopologyTestSuite` tests.

### CODE REVIEW

**Comment by alecco** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2366/

Unrelated

```
[Test Result](https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2366/testReport/) (1 failure / +1)
[Sanity Tests / cdc_test.TestCdc.test_cluster_expansion_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2366/testReport/junit/cdc_test/TestCdc/Sanity_Tests___test_cluster_expansion_with_cdc_Multi_DC_cluster_/)
```

**Comment by nyh** [comment]
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2366/
> 
> Unrelated
> 
> ```
> [Test Result](https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2366/testReport/) (1 failure / +1)
> [Sanity Tests / cdc_test.TestCdc.test_cluster_expansion_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2366/testReport/junit/cdc_test/TestCdc/Sanity_Tests___test_cluster_expansion_with_cdc_Multi_DC_cluster_/)
> ```

Can you please rebase so the test will run with an updated master and maybe fix this spurious failure?

File: test.py [hunk 1]
```diff
@@ -333,7 +334,16 @@ def __init__(self, path, cfg: dict, options: argparse.Namespace, mode: str) -> N
 
         self.create_cluster = self.topology_for_class(topology["class"], topology)
 
-        self.clusters = Pool(cfg.get("pool_size", 2), self.create_cluster)
+        cfg_pool_size = cfg.get("pool_size", 2)
+        # Max 1 scylla server per core
+        cores = psutil.cpu_count(logical=False)
+        rf = topology["replication_factor"]
+        rf_core = int(cores / rf)
+        assert rf_core >= 1, f"Not enough cores to run test with replication factor {rf}"
```
Comment (nyh) [inline test.py:342]: I don't understand why you need to add this assert - it's obviously not fun to run a test with RF=3 on a 2-core antique, but why should we forbid it?
When we have few cores, we should limit unnecessary parallelism (i.e., pool_size), but not forbid necessary parallelism inherent in the test itself.

**Review by nyh — CHANGES_REQUESTED** [review]
I left a request to remove what I think is an unnecessary assert, but beyond that, I think that you and @kostja think again why this "pool_size" configuration parameter still exists after this patch.

After this patch, "pool_size" becomes a a maximum parallelization - so if we have RF=3 and have 30 cores, this code decides it can run 10 tests in parallel - but if "pool_size=4" then it will stop at 4 instead. But why? How does the author of the test suite know it's better to stop at 4 parallel tests instead of 10? It sounds to me like something which isn't useful to have in a configuration file. A command line parameter perhaps (like "make -j<num>"), but I don't see how it is useful in a configuration file.

File: test.py (hunk 1 above)
Comment (alecco) [reply test.py]: Fixed in https://github.com/scylladb/scylladb/commit/146bbd7167d5d67890d32d33c35a440b343f4c02

**Comment by alecco** [comment]
> I left a request to remove what I think is an unnecessary assert, but beyond that, I think that you and @kostja think again why this "pool_size" configuration parameter still exists after this patch.

I mostly agree.

**Comment by alecco** [comment]
Latest is rebased on top of @kbr- 's https://github.com/scylladb/scylladb/pull/11590/

**Comment by alecco** [comment]
Rebased to latest @kbr- 's https://github.com/scylladb/scylladb/pull/11590 https://github.com/scylladb/scylladb/commit/4ddf46015c1df8ab3732296593ec11acacfd0cba

**Comment by kostja** [comment]
@nyh Regarding pool size parameter. yes, it is a heuristic and accurately managing all clusters is difficult. 
Let's begin with the unit tests. Today, in order to evaluate how many unit tests to run concurrently, test.py invokes taskset and creates as many concurrent executions, as there are cores. 
This is already inaccurate. There are unit test which require 2 and 4 cores (--smp parameter to seastar). There are unit tests which are mostly io-bound, and there are others, which are exclusively CPU bound.

Let's add cluster tests to the picture. How many cores does a cluster test use? There is a client and a server in every test now, so it's 2 instead of one. Or it could be a cluster of say, 3 nodes, - in this case, it's 4 instead of one. The good news is that most cluster tests are IO bound - so you can under-provision, i.e. run multiple tests on a single core.

On top of that, let's add the cost of starting the cluster. If a suite is small, starting off too many clusters will be a major cost - so for overall throughput, one should sacrifice concurrency of an individual suite and have a lower number of clusters in the pool. Clusters should be re-used.

Finally, there are tags, filters and build modes. What constitutes the pool of tests test.py needs to run depends on all that. 
So assigning out CPUs is a scheduling task. In order to do it properly, we need to introduce a CPURegistry, similar to HostRegistry, and lease CPUs at start of each test. How many CPUs a test leases would depend on whether the author considers it io-bound or cpu-bound. And even with that it won't answer the question what's the optimal pool size - how many clusters, from concurrency point of view, we need to pre-create, to optimize execution time.

So while we're thinking how to solve it in a general way, I think this patch needs to be queued, since it addresses some present time issues which affect eveyrone.

**Comment by nyh** [comment]
@kostja please note that I never said that accounting resources, or this patch, was not a good idea. What I said was not good, and now said it again, was the per-suite concurrency configuration variable like "4". This is what I suggested we should eliminate. And you're right, we don't have to do that right now.

**Comment by alecco** [comment]
Closing. We should work a better patch later.

**Pull request closed without merging**